README
NAME
    Net::Appliance::Phrasebook - Network appliance command-line phrasebook

VERSION
    version 2.103641

SYNOPSIS
     use Net::Appliance::Phrasebook;
     
 my $pb = Net::Appliance::Phrasebook->new(
         platform => 'IOS',
         source   => '/a/file/somewhere.yml', # optional
     );
     
 print $pb->fetch('a_command_alias'), "\n";

DESCRIPTION
    If you use Perl to manage interactive sessions with with the
    command-line interfaces of networked appliances, then you might find
    this module useful.

    Net::Appliance::Phrasebook is a simple module that contains a number of
    dictionaries for the command-line interfaces of some popular network
    appliances.

    It also supports the use of custom phrasebooks, and of hiearchies of
    dictionaries within phrasebooks.

TERMINOLOGY
    This module is based upon Data::Phrasebook. A *phrasebook* is a file
    which contains one or more dictionaries. A *dictionary* is merely an
    associative array which maps keywords to values. In the case of this
    module, the values happen to be command line interface commands, or
    related data, that help in the remote management of network appliances.

METHODS
  "new"
    This method accepts a list of named arguments (as a hash).

    There is one required named argument, which is the class of device whose
    dictionary you wish to access. The named argument is called "platform".

    One further, optional argument to "new" is the filename of a phrasebook.
    If this is not provided, Net::Appliance::Phrasebook will use its own
    internal phrasebook (see "SUPPORTED SYSTEMS"). This named argument is
    called "source".

    The "new" constructor returns a Data::Phrasebook query object, or
    "undef" on failure.

  "load"
    This is an alias for the "new()" constructor should you prefer to use
    it.

  "fetch"
    Pass this method a single keyword, and it will return the corresponding
    value from the dictionary. It will die on lookup failure, because that's
    what Data::Phrasebook does when there is no successful hit for the given
    keyword in available dictionaries.

SUPPORTED SYSTEMS
    You can select the *platform* that most closely reflects your device.
    There is a hierarchy of platforms, so any entry in a given "lineage"
    will use itself and its "ancestors", in order, for lookups:

     ['FWSM3', 'FWSM', 'PIXOS', 'Cisco']
     ['Aironet', 'IOS', 'Cisco']
     ['PIXOS7', 'Cisco']
     ['CATOS', 'Cisco']
     ['JUNOS', 'Cisco']
     ['HP', 'Cisco']
     ['Nortel', 'Cisco']
     ['ExtremeXOS', 'Cisco']

    For example the value "FWSM" (for Cisco Firewall Services Modules with
    software versions up to 2.x) will fetch commands from the "FWSM"
    dictionary and then the "PIXOS" dictionary, then the "Cisco" dictionary,
    before failing.

CUSTOM PHRASEBOOKS
    Phrasebooks must be written in YAML, with each dictionary being named
    within the top-level associative array in the stream. Please see
    Data::Phrasebook::Loader::YAML for more detail on the format of the
    content of a YAML phrasebook file.

    In the world of network appliances, vendors will sometimes change the
    commands used in or even the appearance of the command line interface.
    This might happen between software version releases, or as a new product
    line is released.

    However, typically there is an ancestry to all these interfaces, so we
    can base a new product's dictionary on an existing dictionary whilst
    overriding some entries with new values. If you study the source to this
    module, you'll see that the bundled phrasebook makes uses of such
    platform families to avoid repetition.

    It is recommended that when creating new phrasebooks you follow this
    pattern. When doing so you must pass an array reference to the
    "platform" argument of "new" and it will be used as a list of
    dictionaries to find entries in, in order. Note that the array reference
    option for the "platform" argument will only work when used with a named
    external source data file.

TIPS
    The phrasebook that ships with this module is stored in a separate file,
    alongside the module itself on your computer. For example:

     .../Net/Appliance/Phrasebook.pm
     .../Net/Appliance/Phrasebook/nas-pb.yml

    So the file you want to copy to start your own phrasebook is
    "nas-pb.yml", as above. Having copied it, make some changes and use that
    file in the "source" named parameter. Make sure you pass the "platform"
    parameter a value too, in that case.

    Read the manual pages for Data::Phrasebook::Loader::YAML and
    Data::Phrasebook to understand what a *default dictionary* is, and why
    you probably always want to have (an empty) one in a phrasebook.

    In YAML, an empty associative array is represented by "{}". Be sure to
    put that into your cutom dictionaries where needed, otherwise
    Data::Phrasebook::Loader::YAML will misbehave.

DIAGNOSTICS
    "missing argument to Net::Appliance::Phrasebook::new"
        You forgot to pass the required "platform" argument to "new".

    "unknown platform: foobar, could not find phrasebook"
        You asked for a dictionary "foobar" that does not exist in the
        internal phrasebook.

    "couldn't find the NAS Phrasebook home directory"
        The module searched for the phrasebook it shipped with, but failed
        to find it. Please report this error (including the message itself)
        to the module maintainer.

    "NAS Phrasebook file at Net/Appliance/Phrasebook/nas-pb.yml does not
    seem to exist"
        The module searched for the phrasebook it shipped with, but failed
        to find it. Please report this error (including the message itself)
        to the module maintainer.

DEPENDENCIES
    Other than the the contents of the standard Perl distribution, you will
    need the following:

    *   Data::Phrasebook::Loader::YAML >= 0.06

    *   Data::Phrasebook >= 0.26

    *   List::MoreUtils

    *   Class::Data::Inheritable

    *   YAML >= 0.62

BUGS
    If you spot a bug or are experiencing difficulties that are not
    explained within the documentation, please send an email to
    oliver@cpan.org or submit a bug to the RT system (http://rt.cpan.org/).
    It would help greatly if you are able to pinpoint problems or even
    supply a patch.

SEE ALSO
    Data::Phrasebook, Net::Appliance::Session,
    Data::Phrasebook::Loader::YAML

AUTHOR
    Oliver Gorwits <oliver@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by University of Oxford.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
{{$NEXT}}

2.103642  2010-12-30 20:31:40 Europe/London

  * Fix error string for PIXOS7
  * Add an ASA device class

2.103641  2010-12-30 19:57:38 Europe/London

  * Add ExtremeXOS phrases courtesy those nice PacketFence people.

2.103640  2010-12-30 19:47:36 Europe/London

  * Port to Dist::Zilla

libnet-appliance-phrasebook-perl (1.8) UNRELEASED; urgency=low

  * Various JUNOS fixes (submitted via forum by quandarypeak)
  * Added PIXOS7 device which has a different pager command (RT#40784)

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 30 Dec 2008 22:38:44 +0000

libnet-appliance-phrasebook-perl (1.7) UNRELEASED; urgency=low

  * Move to Module::Install
  * Update license to be same terms as Perl

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sat,  8 Nov 2008 16:42:09 +0000

libnet-appliance-phrasebook-perl (1.6) UNRELEASED; urgency=low

  * Fix for probably bug in matching configure/privilege mode 

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sun, 19 Oct 2008 01:29:58 +0100

libnet-appliance-phrasebook-perl (1.5) UNRELEASED; urgency=low

  * Character class error for HP and Nortel, spotted by Marc Remijn 
  * Add beginnings of a JUNOS phrasebook, courtesy Eddie Parra

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sun, 19 Oct 2008 00:28:12 +0100

libnet-appliance-phrasebook-perl (1.4) unstable; urgency=low

  * Switch to external YAML phrasebook file instead of a File::Temp and <DATA>

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sat, 23 Aug 2008 12:05:37 +0100

libnet-appliance-phrasebook-perl (1.3) unstable; urgency=low

  * Switch to a File::Temp instead of re-reading <DATA> 

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sun, 17 Aug 2008 18:08:36 +0100

libnet-appliance-phrasebook-perl (1.2) unstable; urgency=low

  * Added phrase for command line completion
  * Bumped version into the real world.

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 03 Jun 2008 17:54:43 +0100

libnet-appliance-phrasebook-perl (0.1) unstable; urgency=low

  * Added 'disconnect' phrase for RT.cpan#35937

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sat, 17 May 2008 20:28:32 +0100

libnet-appliance-phrasebook-perl (0.09) unstable; urgency=low

  * Added Cisco CatOS commands contributed by Jörg Roth

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Thu, 01 May 2008 15:20:00 +0100

libnet-appliance-phrasebook-perl (0.08) unstable; urgency=low

  * Add more error strings

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Fri, 26 Jan 2007 14:58:00 +0000

libnet-appliance-phrasebook-perl (0.07) unstable; urgency=low

  * Added a load of new phrases that were previously hard-coded in
    Net::Appliance::Session
  * Added a 'Cisco' entry to the hierachy to carry these
  * Added a 'CATOS' dictionary, but it's untested and pretty basic
  * Hacked around a bug in Data::Phrasebook whereby a call to dicts() or
    keywords() would fail unless a fetch() had already been done. Now you
    can call keywords() immediately after instantiation
  * Updated tests to cope with changes to built-in dicts
  * Added more info to the POD about custom phrasebooks

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>   Tue, 14 Nov 2006 11:48:00 +0000

libnet-appliance-phrasebook-perl (0.06) unstable; urgency=low

  * Tweaks to the POD

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed, 04 Oct 2006 12:45:00 +0100

libnet-appliance-phrasebook-perl (0.05) unstable; urgency=low

  * Initial Release.

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Mon, 02 Oct 2006 23:40:00 +0100


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Network appliance command-line phrasebook',
  'AUTHOR' => 'Oliver Gorwits <oliver@cpan.org>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Net-Appliance-Phrasebook',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Net::Appliance::Phrasebook',
  'PREREQ_PM' => {
    'Carp' => '0',
    'Class::Data::Inheritable' => '0',
    'Data::Phrasebook' => '0',
    'Data::Phrasebook::Loader::YAML' => '0',
    'File::Basename' => '0',
    'List::MoreUtils' => '0',
    'List::Util' => '0',
    'Symbol' => '0'
  },
  'VERSION' => '2.103641',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Net/Appliance/Phrasebook.pm
package Net::Appliance::Phrasebook;

use strict;
use warnings FATAL => qw(all);

use base qw(Class::Data::Inheritable);

use Data::Phrasebook;
use List::Util qw(first);
use List::MoreUtils qw(after_incl);
use File::Basename;
use Symbol;
use Carp;

__PACKAGE__->mk_classdata('__phrasebook_file');
__PACKAGE__->mk_classdata('__families' => [
    ['FWSM3', 'FWSM', 'PIXOS', 'Cisco'],
    ['ASA', 'PIXOS7', 'PIXOS', 'Cisco'],
    ['Aironet', 'IOS', 'Cisco'],
    ['CATOS', 'Cisco'],
    ['JUNOS', 'Cisco'],
    ['HP', 'Cisco'],
    ['Nortel', 'Cisco'],
    ['ExtremeXOS', 'Cisco'],
]);

sub new {
    my $class = shift;
    my %args  = @_;

    croak "missing argument to Net::Appliance::Phrasebook::new"
        if not defined $args{platform};

    my ($data, $dict);

    # user's own phrasebook
    if (exists $args{source}) {
        $data = $args{source};
        $dict = $args{platform};

    }
    # our internal phrasebook
    else {
        # find and prune down dictionary list
        # NOTE: nested func's from List::* have bad scope, hence using inner grep
        my $family = first { grep { $_ eq $args{platform} } @{$_} } @{__PACKAGE__->__families};
        $dict = [ after_incl { $_ eq $args{platform} } @{$family} ];

        croak "unknown platform: $args{platform}, could not find dictionary"
            if scalar @{$dict} == 0 or ! defined $dict->[0];

        # find our phrasebook file on the filesystem
        (my $pkg_path = __PACKAGE__) =~ s{::}{/}g;
        my (undef, $directory, undef) = fileparse(
            $INC{ $pkg_path .'.pm' }
        );
        croak "couldn't find the NAS Phrasebook home directory"
            if !defined $directory;

        $data = $directory . 'Phrasebook/nas-pb.yml';
        croak "NAS Phrasebook file at $data does not seem to exist"
            if ! -e $data;
    }

    my $self = Data::Phrasebook->new(
        class  => 'Plain',
        loader => 'YAML',
        file   => $data,
        dict   => $dict,
    );
    $self->delimiters(qr{^!}); # it objects to colons
    $self->data('0 but true'); # force load on Phrasebook (a bug in D::P)

    return $self;
}

*{Symbol::qualify_to_ref('load')} = \&new;

1;

# ABSTRACT: Network appliance command-line phrasebook

=head1 SYNOPSIS

 use Net::Appliance::Phrasebook;
 
 my $pb = Net::Appliance::Phrasebook->new(
     platform => 'IOS',
     source   => '/a/file/somewhere.yml', # optional
 );
 
 print $pb->fetch('a_command_alias'), "\n";

=head1 DESCRIPTION

If you use Perl to manage interactive sessions with with the command-line
interfaces of networked appliances, then you might find this module useful.

Net::Appliance::Phrasebook is a simple module that contains a number of
dictionaries for the command-line interfaces of some popular network
appliances.

It also supports the use of custom phrasebooks, and of hiearchies of
dictionaries within phrasebooks.

=head1 TERMINOLOGY

This module is based upon L<Data::Phrasebook>. A I<phrasebook> is a file which
contains one or more dictionaries. A I<dictionary> is merely an associative
array which maps keywords to values. In the case of this module, the values
happen to be command line interface commands, or related data, that help in
the remote management of network appliances.

=head1 METHODS

=head2 C<new>

This method accepts a list of named arguments (as a hash).

There is one required named argument, which is the class of device whose
dictionary you wish to access. The named argument is called C<platform>.

One further, optional argument to C<new> is the filename of a phrasebook. If
this is not provided, Net::Appliance::Phrasebook will use its own internal
phrasebook (see L</"SUPPORTED SYSTEMS">). This named argument is called
C<source>.

The C<new> constructor returns a L<Data::Phrasebook> query object, or C<undef>
on failure.

=head2 C<load>

This is an alias for the C<new()> constructor should you prefer to use it.

=head2 C<fetch>

Pass this method a single keyword, and it will return the corresponding value
from the dictionary. It will die on lookup failure, because that's what
Data::Phrasebook does when there is no successful hit for the given keyword in
available dictionaries.

=head1 SUPPORTED SYSTEMS

You can select the I<platform> that most closely reflects your device. There
is a hierarchy of platforms, so any entry in a given "lineage" will use itself
and its "ancestors", in order, for lookups:

 ['FWSM3', 'FWSM', 'PIXOS', 'Cisco']
 ['ASA', 'PIXOS7', 'PIXOS', 'Cisco']
 ['Aironet', 'IOS', 'Cisco']
 ['CATOS', 'Cisco']
 ['JUNOS', 'Cisco']
 ['HP', 'Cisco']
 ['Nortel', 'Cisco']
 ['ExtremeXOS', 'Cisco']

For example the value C<FWSM> (for Cisco Firewall Services Modules with
software versions up to 2.x) will fetch commands from the C<FWSM> dictionary
and then the C<PIXOS> dictionary, then the C<Cisco> dictionary, before
failing.

=head1 CUSTOM PHRASEBOOKS

Phrasebooks must be written in YAML, with each dictionary being named within
the top-level associative array in the stream. Please see
L<Data::Phrasebook::Loader::YAML> for more detail on the format of the content
of a YAML phrasebook file.

In the world of network appliances, vendors will sometimes change the commands
used in or even the appearance of the command line interface. This might
happen between software version releases, or as a new product line is
released.

However, typically there is an ancestry to all these interfaces, so we can
base a new product's dictionary on an existing dictionary whilst overriding
some entries with new values. If you study the source to this module, you'll
see that the bundled phrasebook makes uses of such platform families to avoid
repetition.

It is recommended that when creating new phrasebooks you follow this pattern.
When doing so you B<must> pass an array reference to the C<platform> argument
of C<new> and it will be used as a list of dictionaries to find entries in, in
order. Note that the array reference option for the C<platform> argument will
only work when used with a named external source data file.

=head1 TIPS

The phrasebook that ships with this module is stored in a separate file,
alongside the module itself on your computer. For example:

 .../Net/Appliance/Phrasebook.pm
 .../Net/Appliance/Phrasebook/nas-pb.yml

So the file you want to copy to start your own phrasebook is C<nas-pb.yml>, as
above. Having copied it, make some changes and use that file in the C<source>
named parameter. Make sure you pass the C<platform> parameter a value too, in
that case.

Read the manual pages for L<Data::Phrasebook::Loader::YAML> and
L<Data::Phrasebook> to understand what a I<default dictionary> is, and why you
probably always want to have (an empty) one in a phrasebook.

In YAML, an empty associative array is represented by C<{}>. Be sure to put
that into your cutom dictionaries where needed, otherwise
Data::Phrasebook::Loader::YAML will misbehave.

=head1 DIAGNOSTICS

=over 4

=item C<missing argument to Net::Appliance::Phrasebook::new>

You forgot to pass the required C<platform> argument to C<new>.

=item C<unknown platform: foobar, could not find phrasebook>

You asked for a dictionary C<foobar> that does not exist in the internal
phrasebook.

=item C<couldn't find the NAS Phrasebook home directory>

The module searched for the phrasebook it shipped with, but failed to find it.
Please report this error (including the message itself) to the module
maintainer.

=item C<NAS Phrasebook file at Net/Appliance/Phrasebook/nas-pb.yml does not seem to exist>

The module searched for the phrasebook it shipped with, but failed to find it.
Please report this error (including the message itself) to the module
maintainer.

=back

=head1 DEPENDENCIES

Other than the the contents of the standard Perl distribution, you will need
the following:

=over 4

=item *

Data::Phrasebook::Loader::YAML >= 0.06

=item *

Data::Phrasebook >= 0.26

=item *

List::MoreUtils

=item *

Class::Data::Inheritable

=item *

YAML >= 0.62

=back

=head1 BUGS

If you spot a bug or are experiencing difficulties that are not explained
within the documentation, please send an email to oliver@cpan.org or submit a
bug to the RT system (http://rt.cpan.org/). It would help greatly if you are
able to pinpoint problems or even supply a patch.

=head1 SEE ALSO

L<Data::Phrasebook>, L<Net::Appliance::Session>,
L<Data::Phrasebook::Loader::YAML>

=cut
